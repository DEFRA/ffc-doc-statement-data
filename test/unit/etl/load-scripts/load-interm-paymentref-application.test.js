const db = require('../../../../app/data')
const { loadIntermPaymentrefApplication } = require('../../../../app/etl/load-scripts/load-interm-paymentref-application')

jest.mock('../../../../app/data', () => ({
  sequelize: {
    query: jest.fn()
  }
}))

describe('loadIntermPaymentrefApplication', () => {
  const startDate = '2023-01-01'
  const transaction = {}

  beforeEach(() => {
    db.sequelize.query.mockClear()
  })

  test('should call sequelize.query with correct SQL and parameters', async () => {
    await loadIntermPaymentrefApplication(startDate, transaction)

    expect(db.sequelize.query).toHaveBeenCalledWith(`
    INSERT INTO etl_interm_paymentref_application(payment_ref, application_id)
    SELECT
      T.payment_ref,
      (SELECT claim_id AS application_id FROM etl_interm_finance_dax D WHERE D.payment_ref = T.payment_ref LIMIT 1)
    FROM etl_interm_total T
    WHERE T.etl_inserted_dt > :startDate;
  `, {
      replacements: {
        startDate
      },
      raw: true,
      transaction
    })
  })

  test('should handle errors thrown by sequelize.query', async () => {
    db.sequelize.query.mockRejectedValue(new Error('Query failed'))

    await expect(loadIntermPaymentrefApplication(startDate, transaction)).rejects.toThrow('Query failed')
  })
})
