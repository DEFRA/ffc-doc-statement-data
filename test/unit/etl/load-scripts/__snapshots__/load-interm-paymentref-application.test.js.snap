// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadIntermPaymentrefApplication should call sequelize.query with correct SQL and parameters 1`] = `
[MockFunction] {
  "calls": [
    [
      "
    INSERT INTO public."etlIntermPaymentrefApplication"("paymentRef", "applicationId")
    SELECT
      T."paymentRef",
      (SELECT "claimId" AS "applicationId" FROM public."etlIntermFinanceDax" D WHERE D."paymentRef" = T."paymentRef" LIMIT 1)
    FROM public."etlIntermTotal" T
    WHERE T."etlInsertedDt" > :startDate
    ON CONFLICT ("paymentRef", "applicationId") DO NOTHING;
  ",
      {
        "raw": true,
        "replacements": {
          "startDate": "2023-01-01",
        },
        "transaction": {},
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
